src/algebraic_geometry/projective_spectrum/structure_sheaf.lean

```diff
@@ -0,0 +1,545 @@
+
+@[simp] lemma deg_zero : (0 : num_denom_same_deg x).deg = 0 := rfl
+@[simp] lemma num_zero : (0 : num_denom_same_deg x).num = 0 := rfl
+@[simp] lemma denom_zero : ((0 : num_denom_same_deg x).denom : A) = 1 := rfl
+
+instance : has_mul (num_denom_same_deg x) :=
+{ mul := λ p q,
+  { deg := p.deg + q.deg,
+    num := ⟨p.num * q.num, mul_mem p.num.prop q.num.prop⟩,
+    denom := ⟨p.denom * q.denom, mul_mem p.denom.prop q.denom.prop⟩,
+    denom_not_mem := λ r, or.elim (x.is_prime.mem_or_mem r) p.denom_not_mem q.denom_not_mem } }
+
+lemma deg_mul (c1 c2 : num_denom_same_deg x) : (c1 * c2).deg = c1.deg + c2.deg := rfl
+lemma num_mul (c1 c2 : num_denom_same_deg x) : ((c1 * c2).num : A) = c1.num * c2.num := rfl
+lemma denom_mul (c1 c2 : num_denom_same_deg x) : ((c1 * c2).denom : A) = c1.denom * c2.denom := rfl
+
+instance : has_add (num_denom_same_deg x) :=
+{ add := λ c1 c2,
+  { deg := c1.deg + c2.deg,
+    num := ⟨c1.denom.1 * c2.num.1 + c2.denom.1 * c1.num.1,
+      add_mem _ (mul_mem c1.denom.2 c2.num.2)
+        (add_comm c2.deg c1.deg ▸ mul_mem c2.denom.2 c1.num.2)⟩,
+    denom := ⟨c1.denom.1 * c2.denom.1, mul_mem c1.denom.2 c2.denom.2⟩,
+    denom_not_mem := λ r, or.elim (x.is_prime.mem_or_mem r) c1.denom_not_mem c2.denom_not_mem } }
+
+lemma deg_add (c1 c2 : num_denom_same_deg x) : (c1 + c2).deg = c1.deg + c2.deg := rfl
+lemma num_add (c1 c2 : num_denom_same_deg x) :
+  ((c1 + c2).num : A) = c1.denom * c2.num + c2.denom * c1.num := rfl
+lemma denom_add (c1 c2 : num_denom_same_deg x) :
+  ((c1 + c2).denom : A) = c1.denom * c2.denom := rfl
+
+instance : has_neg (num_denom_same_deg x) :=
+{ neg := λ c, ⟨c.deg, ⟨-c.num.1, neg_mem _ c.num.2⟩, c.denom, c.denom_not_mem⟩ }
+
+lemma deg_neg (c : num_denom_same_deg x) : (-c).deg = c.deg := rfl
+lemma num_neg (c : num_denom_same_deg x) : (-c).num.1 = -c.num.1 := rfl
+lemma denom_neg (c : num_denom_same_deg x) : (-c).denom.1 = c.denom.1 := rfl
+
+instance : comm_monoid (num_denom_same_deg x) :=
+{ one := 1,
+  mul := (*),
+  mul_assoc := λ c1 c2 c3, ext _ (add_assoc _ _ _) (mul_assoc _ _ _) (mul_assoc _ _ _),
+  one_mul := λ c, ext _ (zero_add _) (one_mul _) (one_mul _),
+  mul_one := λ c, ext _ (add_zero _) (mul_one _) (mul_one _),
+  mul_comm := λ c1 c2, ext _ (add_comm _ _) (mul_comm _ _) (mul_comm _ _) }
+
+def num_denom_same_deg.embedding (p : num_denom_same_deg x) : at x :=
+localization.mk p.num ⟨p.denom, p.denom_not_mem⟩
+
+def homogeneous_localization : Type* := quotient (setoid.ker $ num_denom_same_deg.embedding x)
+
+variable {x}
+def homogeneous_localization.val (y : homogeneous_localization x) : at x :=
+quotient.lift_on' y (num_denom_same_deg.embedding x) $ λ _ _, id
+
+variable (x)
+lemma homogeneous_localization.val_injective :
+  function.injective (@homogeneous_localization.val _ _ _ _ _ 𝒜 _ x) :=
+λ a b, quotient.rec_on_subsingleton₂' a b $ λ a b h, quotient.sound' h
+
+instance : has_pow (homogeneous_localization x) ℕ :=
+{ pow := λ z n, quotient.lift_on' z
```

Comment:
`quotient.map'` might be a little tidier here

Reply:
Indeed.